processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -658,37 +671,42 @@ public DateTime getMaxIngestedEventTime()
     return maxIngestedEventTime;
   }
 
-  class DimensionHolder
+  private class DimensionHolder
   {
-    private final Map<String, DimDim> dimensions;
+    private DimDim[] dimensions;
 
-    DimensionHolder()
+    private DimensionHolder(int numDimension)
     {
-      dimensions = Maps.newConcurrentMap();
+      dimensions = new DimDim[Math.max(32, numDimension)];
```

Comment:
also, the previous data structure was a concurrent map, how is locking done over this array?